batch/src/main/java/is/hail/JVMEntryway.java

```diff
@@ -0,0 +1,222 @@
+package is.hail;
+
+import java.io.*;
+import java.lang.reflect.*;
+import java.net.*;
+import java.nio.*;
+import java.nio.charset.*;
+import java.util.*;
+import java.util.concurrent.*;
+import org.newsclub.net.unix.*;
+
+class JVMEntryway {
+  public static String throwableToString(Throwable t) throws IOException {
+    try (StringWriter sw = new StringWriter();
+         PrintWriter pw = new PrintWriter(sw)) {
+      t.printStackTrace(pw);
+      return sw.toString();
+    }
+  }
+
+  private static int FINISH_USER_EXCEPTION = 0;
+  private static int FINISH_ENTRYWAY_EXCEPTION = 1;
+  private static int FINISH_NORMAL = 2;
+  private static int FINISH_CANCELLED = 3;
+
```

Comment:
There's no JVM_EOS (4) here, but it's in the worker.py code.

Reply:
I added it here with a note that it should never be used here because the JVM cannot know that is unexpectedly sending EOS.